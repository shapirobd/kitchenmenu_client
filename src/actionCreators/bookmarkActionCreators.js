import axios from "axios";
import { BOOKMARK_RECIPE, UNBOOKMARK_RECIPE } from "../components/actionTypes";

// adds bookmarked recipe to database as well as redux state
// by dispatching action created by bookmarkedRecipe()
export const bookmarkRecipe = (username, recipeId) => {
	return async (dispatch) => {
		try {
			await axios.post(
				`${
					process.env.REACT_APP_API_URL || "http://localhost:5000"
				}/users/bookmarkRecipe`,
				{
					username,
					recipeId,
				}
			);
			dispatch(bookmarkedRecipe(recipeId));
		} catch (e) {
			console.error(e);
		}
	};
};

// returns action to be dispatched containing id of recipe to be bookmarked
const bookmarkedRecipe = (recipeId) => {
	return {
		type: BOOKMARK_RECIPE,
		payload: {
			recipeId,
		},
	};
};

// removes bookmarked recipe from database as well as redux state
// by dispatching action created by unbookmarkedRecipe()
export const unbookmarkRecipe = (username, recipeId) => {
	return async (dispatch) => {
		try {
			await axios.post(
				`${
					process.env.REACT_APP_API_URL || "http://localhost:5000"
				}/users/unbookmarkRecipe`,
				{
					username,
					recipeId,
				}
			);
			dispatch(unbookmarkedRecipe(recipeId));
		} catch (e) {
			console.error(e);
		}
	};
};

// returns action to be dispatched containing id of recipe to be unbookmarked
const unbookmarkedRecipe = (recipeId) => {
	return {
		type: UNBOOKMARK_RECIPE,
		payload: {
			recipeId,
		},
	};
};
