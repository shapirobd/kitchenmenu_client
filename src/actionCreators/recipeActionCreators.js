import axios from "axios";
import { LOAD_FEED, LOAD_RECIPE, FILTER_FEED } from "../components/actionTypes";
import createMacrosParams from "../helpers/createMacrosParams";

// retrieves all recipes from spoonacular API (paginated 40 per page) and updates
// the feed & page number in redux state by dispatching array of recipes and the page number
export const loadFeed = (page = 1, filterData = { diets: [], macros: {} }) => {
	return async (dispatch) => {
		try {
			const { diets, macros, recipeName } = filterData;
			const macrosParams = createMacrosParams(macros);
			const dietsParams = diets.join(",");
			const recipes = await axios.get(
				"https://api.spoonacular.com/recipes/complexSearch",
				{
					params: {
						apiKey: "73baf9bb95a14f5fb4d71e2f12ab8479",
						offset: 40 * (page - 1),
						number: 40,
						diet: dietsParams,
						...macrosParams,
						query: recipeName,
					},
				}
			);
			dispatch(loadedFeed(recipes.data, page));
		} catch (e) {
			console.error(e);
		}
	};
};

// returns action containing recipes to include in redux state's feed,
// total number of results found to set as redux state's totalResults,
// and the current page number to set as redux state's page
const loadedFeed = (data, page) => {
	return {
		type: LOAD_FEED,
		payload: {
			recipes: data.results,
			totalResults: data.totalResults <= 900 ? data.totalResults : 900,
			page,
		},
	};
};

// retrieves data on a single recipe from spoonacular API and updates
// currentRecipe in redux state by dispatching recipe info & instructions info
export const loadRecipe = (recipeId) => {
	console.log("2: ", recipeId);
	return async (dispatch) => {
		try {
			const recipe = await axios.get(
				`https://api.spoonacular.com/recipes/${recipeId}/information`,
				{
					params: {
						apiKey: "73baf9bb95a14f5fb4d71e2f12ab8479",
						includeNutrition: true,
					},
				}
			);
			const instructions = await axios.get(
				`https://api.spoonacular.com/recipes/${recipeId}/analyzedInstructions`,
				{
					params: {
						apiKey: "73baf9bb95a14f5fb4d71e2f12ab8479",
						stepBreakdown: true,
					},
				}
			);
			dispatch(loadedRecipe(recipe.data, instructions.data[0]));
		} catch (e) {
			console.error(e);
		}
	};
};

// returns action containing information to include in redux state's currentRecipe
const loadedRecipe = (recipe, instructions) => {
	return {
		type: LOAD_RECIPE,
		payload: { currentRecipe: { recipe, instructions } },
	};
};

// retrieves all recipes from spoonacular API with given ingredients (pagination not an option)
// and updates feed, totalResults, page, and isIngredientBased in redux state by
// dispatching array of found recipes and the length of that array
export const getRecipesByIngredients = (ingredients) => {
	return async (dispatch) => {
		try {
			const ingredientsParams = ingredients.join(",");
			const recipes = await axios.get(
				"https://api.spoonacular.com/recipes/findByIngredients",
				{
					params: {
						apiKey: "73baf9bb95a14f5fb4d71e2f12ab8479",
						offset: 0,
						number: 6900,
						ingredients: ingredientsParams,
					},
				}
			);
			dispatch(
				filteredFeed({ results: recipes.data, totalResults: recipes.length })
			);
		} catch (e) {
			console.error(e);
		}
	};
};

// export const getRecipesByKeywords = (keywords) => {
// 	return async (dispatch) => {
// 		try {
// 			const ingredientsParams = ingredients.join(",");
// 			const recipes = await axios.get(
// 				"https://api.spoonacular.com/recipes/findByIngredients",
// 				{
// 					params: {
// 						apiKey: "73baf9bb95a14f5fb4d71e2f12ab8479",
// 						offset: 0,
// 						number: 6900,
// 						query: keywords,
// 					},
// 				}
// 			);
// 			dispatch(
// 				filteredFeed({ results: recipes.data, totalResults: recipes.length })
// 			);
// 		} catch (e) {
// 			console.error(e);
// 		}
// 	};
// };

// const gotRecipesByIngredients = (results) => {
// 	return {
// 		type: FILTER_BY_INGREDIENTS,
// 		payload: { recipes: data.results, totalResults: data.totalResults },
// 	};
// };

// export const filterFeed = (filterData, page = 1) => {
// 	return async (dispatch) => {
// 		try {
// 			const { diets, macros, recipeName } = filterData;
// 			const macrosParams = createMacrosParams(macros);
// 			const dietsParams = diets.join(",");
// 			const recipes = await axios.get(
// 				"https://api.spoonacular.com/recipes/complexSearch",
// 				{
// 					params: {
// 						apiKey: "73baf9bb95a14f5fb4d71e2f12ab8479",
// 						offset: 40 * (page - 1),
// 						number: 40,
// 						diet: dietsParams,
// 						...macrosParams,
// 						query: recipeName,
// 					},
// 				}
// 			);
// 			console.log(recipes.data);
// 			dispatch(filteredFeed(recipes.data, page));
// 		} catch (e) {
// 			console.error(e);
// 		}
// 	};
// };

// returns action containing information to include in redux state's feed, totalResults, and page
const filteredFeed = (data, page) => {
	return {
		type: FILTER_FEED,
		payload: { recipes: data.results, totalResults: data.totalResults, page },
	};
};
