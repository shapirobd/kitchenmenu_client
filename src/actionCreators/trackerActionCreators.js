import axios from "axios";
import { ADD_EATEN_MEAL, REMOVE_EATEN_MEAL } from "../components/actionTypes";
import { API_URL } from "../constants";

// const API_URL = "http://localhost:5000";

// adds eaten meal to database as well as redux state for a given date
// by dispatching action created by addedEatenMeal()
export const addEatenMeal = (username, recipeId, date, nutrients) => {
	return async (dispatch) => {
		try {
			const macros = {
				calories: Math.round(nutrients[0].amount),
				fat: Math.round(nutrients[1].amount),
				carbs: Math.round(nutrients[3].amount),
				protein: Math.round(nutrients[8].amount),
			};
			await axios.post(`${API_URL}/users/addEatenMeal`, {
				username,
				recipeId,
				date,
				nutrients: macros,
			});
			dispatch(addedEatenMeal(recipeId, date, macros));
		} catch (e) {
			console.error(e);
		}
	};
};

// returns action to be dispatched containing id of recipe
// to be eaten and date that it was eaten
const addedEatenMeal = (recipeId, date, nutrients) => {
	const { calories, fat, carbs, protein } = nutrients;
	return {
		type: ADD_EATEN_MEAL,
		payload: {
			date,
			meal: {
				id: recipeId,
				calories,
				fat,
				carbs,
				protein,
			},
		},
	};
};

// removes eaten meal from database as well as redux state for a given date
// by dispatching action created by removedEatenMeal()
export const removeEatenMeal = (username, recipeId, date) => {
	return async (dispatch) => {
		try {
			await axios.post(`${API_URL}/users/removeEatenMeal`, {
				username,
				recipeId,
				date,
			});
			dispatch(removedEatenMeal(recipeId, date));
		} catch (e) {
			console.error(e);
		}
	};
};

// returns action to be dispatched containing id of recipe
// to be removed and date that it was eaten
const removedEatenMeal = (recipeId, date) => {
	return {
		type: REMOVE_EATEN_MEAL,
		payload: {
			recipeId,
			date,
		},
	};
};
