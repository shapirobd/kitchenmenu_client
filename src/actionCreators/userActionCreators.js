import axios from "axios";
import { LOGIN, LOGOUT } from "../components/actionTypes";

// adds user to database as well as redux state
// by dispatching action created by registered()
export const register = (data) => {
	return async (dispatch) => {
		try {
			const resp = await axios.post(
				`${
					process.env.REACT_APP_API_URL || "http://localhost:5000"
				}/auth/register`,
				data
			);
			// const userResp = await axios.get(`${API_URL}/users/${data.username}`);
			const { user, token } = resp.data;
			dispatch(registered(token, user));
		} catch (e) {
			console.error(e);
		}
	};
};

// returns action to be dispatched containing information on the
// newly registered user
const registered = (token, user) => {
	return {
		type: LOGIN,
		payload: {
			user,
			token,
		},
	};
};

// verifies that username & password from data are valid
// if valid, adds the user to redux state by dispatching action created by loggedIn()
// in invalid, throws error
export const login = (data) => {
	return async (dispatch) => {
		try {
			const resp = await axios.post(
				`${
					process.env.REACT_APP_API_URL || "http://localhost:5000"
				}/auth/login`,
				data
			);
			const { user, token } = resp.data;
			dispatch(loggedIn(token, user));
		} catch (e) {
			console.error(e);
		}
	};
};

// returns action to be dispatched containing information on the
// logged in user
const loggedIn = (token, user) => {
	return {
		type: LOGIN,
		payload: {
			user,
			token,
		},
	};
};

// returns action to be dispatched that ends up removing
// the user's information from redux state
export const logout = () => {
	return {
		type: LOGOUT,
	};
};

// edits the information on a given user in the database
// as well as redux state by dispatching action created by edittedProfile()
export const editProfile = (username, data) => {
	return async (dispatch) => {
		try {
			const editResp = await axios.patch(
				`${process.env.API_URL || "http://localhost:5000"}/users/${username}`,
				data
			);
			dispatch(edittedProfile(editResp.data));
		} catch (e) {
			console.error(e);
		}
	};
};

// returns action to be dispatched containing information on the
// logged in user
const edittedProfile = (user) => {
	return {
		type: "EDIT_PROFILE",
		payload: { user },
	};
};
