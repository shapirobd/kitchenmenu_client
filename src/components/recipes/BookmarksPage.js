import React, { useState, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import axios from "axios";
import RecipeGrid from "./RecipeGrid";
import { Typography } from "@material-ui/core";
import { useStyles } from "./styles/BookmarksPageStyles";
import useWindowDimensions from "../../customHooks/getWindowDimensions";
import { unbookmarkRecipe } from "../../actionCreators/bookmarkActionCreators";

const BookmarksPage = () => {
	const classes = useStyles();
	const dispatch = useDispatch();
	const user = useSelector((state) => state.user);
	const bookmarks = useSelector((state) => state.user.bookmarks);
	const [fullBookmarks, setFullBookmarks] = useState([]);
	const { height, width } = useWindowDimensions();

	const removeBookmark = (id) => {
		setFullBookmarks(fullBookmarks.filter((bookmark) => bookmark.id !== id));
		dispatch(unbookmarkRecipe(user.username, id));
	};

	useEffect(() => {
		const loadBookmarks = async () => {
			const resp = await axios.get(
				`https://api.spoonacular.com/recipes/informationBulk`,
				{
					params: {
						ids: bookmarks.join(","),
						apiKey: "73baf9bb95a14f5fb4d71e2f12ab8479",
					},
				}
			);
			setFullBookmarks(resp.data);
		};
		if (bookmarks.length) loadBookmarks();
	}, [bookmarks]);

	return (
		<>
			<div className={width > 599 ? classes.root : classes.mobileRoot}>
				<div
					style={
						width > 599
							? { width: `${width - 240}px`, height: `${height}px` }
							: { width: "100%", height: "100%" }
					}
				>
					<div
						style={{
							width: "100%",
							backgroundColor: "white",
							display: "flex",
							flexDirection: "column",
							justifyContent: "space-around",
							alignItems: "center",
							borderRadius: "5px",
							margin: "0 0 10px 0",
							paddingBottom: "30px",
						}}
					>
						<Typography variant="h4" style={{ margin: "20px" }}>
							Bookmarks
						</Typography>
						<RecipeGrid
							feed={fullBookmarks}
							areBookmarks={true}
							removeBookmark={removeBookmark}
						/>
					</div>
				</div>
			</div>
		</>
	);
};

export default BookmarksPage;
