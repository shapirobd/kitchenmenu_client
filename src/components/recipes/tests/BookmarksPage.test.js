import React from "react";
import { render, screen } from "@testing-library/react";
import BookmarksPage from "../BookmarksPage";
import { MemoryRouter } from "react-router-dom";
import { PersistGate } from "redux-persist/integration/react";
import { Provider } from "react-redux";
import { store, persistor } from "../../../configureStore";
import { login } from "../../../actionCreators/userActionCreators";

beforeAll(async () => {
	await store.dispatch(
		login({ username: "shapirobd", password: "Pot3ntiat321!" })
	);
});

it("should render without crashing", () => {
	render(
		<Provider store={store}>
			<MemoryRouter>
				<PersistGate loading={null} persistor={persistor}>
					<BookmarksPage />
				</PersistGate>
			</MemoryRouter>
		</Provider>
	);
});

it("should match screenshot", () => {
	const { asFragment } = render(
		<Provider store={store}>
			<MemoryRouter>
				<PersistGate loading={null} persistor={persistor}>
					<BookmarksPage />
				</PersistGate>
			</MemoryRouter>
		</Provider>
	);

	expect(asFragment()).toMatchSnapshot();
});
