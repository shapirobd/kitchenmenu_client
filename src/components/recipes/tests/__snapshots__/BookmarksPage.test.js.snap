// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match screenshot 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-13"
  >
    <div
      style="width: 784px; height: 768px;"
    >
      <div
        style="width: 100%; background-color: white; display: flex; flex-direction: column; justify-content: space-around; align-items: center; border-radius: 5px; margin: 0px 0px 10px 0px; padding-bottom: 30px;"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4"
          style="margin: 20px;"
        >
          Bookmarks
        </h4>
        <div
          class="makeStyles-root-15"
        >
          <ul
            class="MuiGridList-root makeStyles-gridList-16"
            style="margin: -2px;"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
